File: src/main/java/ir/alireza/sadeghi/shiraz/BoardMethods.java
package ir.alireza.sadeghi.shiraz;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.Hashtable;

/*
 * Has methods which with you can:
 * -clone a hash board
 * -compare two boards
 * -get the marbles in one board from a certain player
 */

public class BoardMethods {
	public static ArrayList<String> hash = Board.hash;
	private final static Logger logger = LogManager.getLogger(BoardMethods.class);
	//copies a whole hash board - depends on the one you get as input
	public static Hashtable<String, Hexagon> copyHashBoard(Hashtable<String, Hexagon> hex){
		logger.traceEntry();
		logger.trace("hash size " + hash.size());
		Hashtable<String, Hexagon> newBoard = new Hashtable();
		for(int i = 0; hash.size() > i; i++) {
			newBoard.put(hash.get(i), hex.get(hash.get(i)).deepClone());
		}
		return newBoard;
	}
	
	//compares two hashTables, if they contain the same marbles (from the same player), at the same spot, then it returns true
	public static boolean compareHashTables(Hashtable<String, Hexagon> one, Hashtable<String, Hexagon> two) {
		for(int i = 0; i < hash.size(); i++) {
			if(!one.get(hash.get(i)).empty && !two.get(hash.get(i)).empty) {
				if (one.get(hash.get(i)).marble.playerNumber != two.get(hash.get(i)).marble.playerNumber) {
					return false;
				}
			}
			else if((one.get(hash.get(i)).empty && !two.get(hash.get(i)).empty) || (!one.get(hash.get(i)).empty && two.get(hash.get(i)).empty)) {
				return false;
			}
		}
		return true;
	}
	
// creates and returns an array list with all the marbles from a certain player
//	public static ArrayList<String> getMarblesPlayer(Hashtable<String, Hexagon> board, int playerNumber){
//		ArrayList<String> marbles = new ArrayList<>();
//		for (int i = 0; i < hash.size(); i++) {
//			if (!board.get(hash.get(i)).empty) {
//				if (board.get(hash.get(i)).marble.playerNumber == playerNumber) {
//					marbles.add(hash.get(i));
//				}
//			}
//		}
//		return marbles;
//	}
	
	//repetition checker - return false if it is a repetition
	public static boolean repetitionChecker(Hashtable<String, Hexagon> current) {
		ArrayList<Hashtable<String, Hexagon>> tb = GameData.tb.getTB();
		for (int i = 0; i < tb.size(); i++) {
			if (compareHashTables(tb.get(i), current)) {
				return false;
			}
		}
		return true;
	}
	
	//repetition checker - return false if it is a repetition
	public static boolean moveRepetitionChecker(String first, String second, String third, int playerNumber, int direction) {
		ArrayList<String> check = null;
		int olddirection = 0;
		if(playerNumber ==1) {
			check = GameData.tb.oldMovePlayer1;
			olddirection = GameData.tb.direct1;
		}
		else if(playerNumber ==2) {
			check = GameData.tb.oldMovePlayer2;
			olddirection = GameData.tb.direct2;
		}
		else {
			check = GameData.tb.oldMovePlayer3;
			olddirection = GameData.tb.direct3;
		}

		if (first != null && second == null && third == null) {
			if (check.size() == 1) {
				if (check.get(0) == first && GameData.rows.oppositeDirection(olddirection) == direction) {
					return false;
				}
			}
		}
		
		if (first != null && second != null && third == null) {
			if (check.size()== 2) {
				if ((check.get(0) == second && check.get(1) == first || check.get(0) == first && check.get(1) == second) && direction == GameData.rows.oppositeDirection(olddirection)) {
					return false;
				}
			}
			
		}
		if (first != null && second != null && third != null) {
			if(check.size() == 3) {
				if (check.contains(first) && check.contains(second) && check.contains(third) && direction == GameData.rows.oppositeDirection(olddirection)) {
					return false;
				}
			}
		}
		return true;
	}
}

File: src/main/java/ir/alireza/sadeghi/shiraz/BoardMethods.kt
package ir.alireza.sadeghi.shiraz

import org.apache.logging.log4j.LogManager
import java.util.*

/*
 * Has methods which with you can:
 * -clone a hash board
 * -compare two boards
 * -get the marbles in one board from a certain player
 */
object BoardMethods {
    var hash: ArrayList<String> = Board.hash
    private val logger = LogManager.getLogger(BoardMethods::class.java)

    //copies a whole hash board - depends on the one you get as input
    fun copyHashBoard(hex: Hashtable<String?, Hexagon?>?): Hashtable<String, Hexagon> {
        logger.traceEntry()
        logger.trace("hash size " + hash.size)
        val newBoard: Hashtable<String, Hexagon> = Hashtable()
        var i = 0
        while (hash.size > i) {
            newBoard[hash[i]] = hex!![hash[i]]!!.deepClone()
            i++
        }
        return newBoard
    }

    //compares two hashTables, if they contain the same marbles (from the same player), at the same spot, then it returns true
    fun compareHashTables(one: Hashtable<String?, Hexagon?>?, two: Hashtable<String?, Hexagon?>?): Boolean {
        for (i in hash.indices) {
            if (!one!![hash[i]]!!.empty && !two!![hash[i]]!!.empty) {
                if (one[hash[i]]!!.marble?.playerNumber != two[hash[i]]!!.marble?.playerNumber) {
                    return false
                }
            } else if (one[hash[i]]!!.empty && !two!![hash[i]]!!.empty || !one[hash[i]]!!.empty && two!![hash[i]]!!.empty) {
                return false
            }
        }
        return true
    }

    // creates and returns an array list with all the marbles from a certain player
    //	public static ArrayList<String> getMarblesPlayer(Hashtable<String, Hexagon> board, int playerNumber){
    //		ArrayList<String> marbles = new ArrayList<>();
    //		for (int i = 0; i < hash.size(); i++) {
    //			if (!board.get(hash.get(i)).empty) {
    //				if (board.get(hash.get(i)).marble.playerNumber == playerNumber) {
    //					marbles.add(hash.get(i));
    //				}
    //			}
    //		}
    //		return marbles;
    //	}
    //repetition checker - return false if it is a repetition
    fun repetitionChecker(current: Hashtable<String?, Hexagon?>?): Boolean {
        val tb = GameData.tb.tB as ArrayList<Hashtable<String?, Hexagon?>>
        for (i in tb.indices) {
            if (compareHashTables(tb[i], current)) {
                return false
            }
        }
        return true
    }

    //repetition checker - return false if it is a repetition
    fun moveRepetitionChecker(
        first: String?,
        second: String?,
        third: String?,
        playerNumber: Int,
        direction: Int
    ): Boolean {
        var check: ArrayList<String>?
        var oldDirection = 0
        when (playerNumber) {
            1 -> {
                check = Traceback.oldMovePlayer1
                oldDirection = Traceback.direct1
            }
            2 -> {
                check = Traceback.oldMovePlayer2
                oldDirection = Traceback.direct2
            }
            else -> {
                check = Traceback.oldMovePlayer3
                oldDirection = Traceback.direct3
            }
        }
        if (first != null && second == null && third == null) {
            if (check?.size == 1) {
                if (check[0] === first && GameData.rows?.oppositeDirection(oldDirection) == direction) {
                    return false
                }
            }
        }
        if (first != null && second != null && third == null) {
            if (check?.size == 2) {
                if ((check[0] === second && check[1]!! === first || check[0] === first && check[1] === second) && direction == GameData.rows?.oppositeDirection(
                        oldDirection
                    )
                ) {
                    return false
                }
            }
        }
        if (first != null && second != null && third != null) {
            if (check?.size == 3) {
                if (check.contains(first) && check.contains(second) && check.contains(third) && direction == GameData.rows?.oppositeDirection(
                        oldDirection
                    )
                ) {
                    return false
                }
            }
        }
        return true
    }
}